GreedyParameterPointSelector.C

```diff
@@ -0,0 +1,465 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: This class greedily selects parameter points
+//              for the construction of a ROM database.
+
+#include "GreedyParameterPointSelector.h"
+#include "Vector.h"
+#include "mpi.h"
+#include <cmath>
+#include <limits.h>
+#include <fstream>
+
+namespace CAROM {
+
+GreedyParameterPointSelector::GreedyParameterPointSelector(
+    std::vector<Vector> parameter_points,
+    double tolerance,
+    double saturation,
+    int subset_size,
+    int convergence_subset_size,
+    bool use_centroid,
+    int random_seed,
+    bool debug_algorithm)
+{
+
+    constructObject(parameter_points, tolerance, saturation,
+        subset_size, convergence_subset_size, use_centroid, random_seed, debug_algorithm);
+}
+
+GreedyParameterPointSelector::GreedyParameterPointSelector(
+    std::vector<double> parameter_points,
+    double tolerance,
+    double saturation,
+    int subset_size,
+    int convergence_subset_size,
+    bool use_centroid,
+    int random_seed,
+    bool debug_algorithm)
+{
+    //convert parameter_points from double to Vector
+    CAROM_VERIFY(parameter_points.size() > 0);
+
+    std::vector<Vector> parameter_points_vec;
+    for (int i = 0; i < parameter_points.size(); i++) {
+        Vector vec(1, false);
+        vec.item(0) = parameter_points[i];
+        parameter_points_vec.push_back(vec);
+    }
+
+    constructObject(parameter_points_vec, tolerance, saturation,
+        subset_size, convergence_subset_size, use_centroid, random_seed, debug_algorithm);
+}
+
+GreedyParameterPointSelector::GreedyParameterPointSelector(
+    double param_space_min,
+    double param_space_max,
+    double param_space_size,
+    double tolerance,
+    double saturation,
+    int subset_size,
+    int convergence_subset_size,
+    bool use_centroid,
+    int random_seed,
+    bool debug_algorithm)
+{
+    //convert parameter_points from double to Vector
+    CAROM_VERIFY(param_space_max > param_space_min);
+    CAROM_VERIFY(param_space_size > 0);
+
+    std::vector<Vector> parameter_points_vec;
+    Vector vec(1, false);
+    if (param_space_size == 1)
```

Comment:
Why are the values 1 and 2 for `param_space_size` treated specially? It will not be clear to the user, and should at least be documented. It would be simpler just to use the else case only, with a positive integer value for `param_space_size`

Reply:
There's a strange case where if a user gives a range, but only with a size of 1, using the else will return just the bottom value. Similary for 2, it will return the bottom and top value, and nothing in the middle. Maybe there's a better segmenting equation that can do the right thing for all cases.

Reply:
Yes, those are some tricky cases. I suppose we should try to provide the simplest interface for the user to understand. I'm not sure what the user would have in mind when specifying 1. If they really want 1 point, they could set the min and max to be the same. In the case of 2 points, the min and max values make sense (the else case). Just using the else case makes sense to me in general, provided it is documented.

Reply:
What you implemented so far was to use the specified number of points in the interior (excluding endpoints) for the cases 1 and 2, but including the endpoints for cases 3 and higher. Maybe a `bool` to specify whether to include endpoints would be more general and clearer. Documenting the case of "endpoints=true" and "size=1 or 2" would also be helpful.